namd/src/GlobalMaster.h

```diff
@@ -0,0 +1,205 @@
+   the positions of atoms and groups, and provides a list of requested
+   atoms, forces, and groups in return.
+
+   I'm not going to do groups for now, because they were done badly
+   originally.  A better solution is necessary.  Hint: multiple
+   groups are required.
+
+   The expected usage (by ComputeGlobalMasterServer) is:
+   1) receive and store the data message from all nodes
+   2) call processData with that data
+   3) collect new forces and atoms
+   4) send them as a message to all the nodes
+   Repeat until done.
+*/
+
+#ifndef GLOBALMASTER_H
+#define GLOBALMASTER_H
+
+#include "NamdTypes.h"
+class Lattice;
+class SubmitReduction;
+
+class GlobalMaster {
+ public:
+
+  /* This passes the atom and group data to the master, which then
+     performs any necessary calculations and updates its results. */
+  void processData(AtomIDList::iterator a_i,
+		   AtomIDList::iterator a_e,
+		   PositionList::iterator p_i,
+		   PositionList::iterator g_i,
+		   PositionList::iterator g_e,
+		   BigRealList::iterator gm_i,
+		   BigRealList::iterator gm_e,
+		   ForceList::iterator gtf_i,
+		   ForceList::iterator gtf_e,
+                   IntList::iterator goi_i,
+                   IntList::iterator goi_e,
+                   BigRealList::iterator gov_i,
+                   BigRealList::iterator gov_e,
+		   AtomIDList::iterator last_atoms_forced_i,
+		   AtomIDList::iterator last_atoms_forced_e,
+		   ForceList::iterator last_forces_i,
+		   AtomIDList::iterator,
+		   AtomIDList::iterator,
+		   ForceList::iterator);
+
+  int step;  // set by server to current timestep before processData
+  int globalMasterStep;  // set by server to current timestep / globalMasterFreq before processData
+  int old_num_groups_requested;  // used for group forces
+
+  bool changedAtoms(); // false if the atom IDs haven't changed
+  const AtomIDList &requestedAtoms(); // the atom ids requested
+  bool changedForces(); // false if the forces haven't changed
+  const AtomIDList &forcedAtoms(); // the atoms the forces are applied to
+  const ForceList &appliedForces(); // the corresponding applied forces
+  bool changedGroups(); // false if the groups haven't changed
+  const ResizeArray<AtomIDList> &requestedGroups(); // the requested groups
+  const ForceList &groupForces(); // the corresponding forces on groups
+  bool changedGridObjs(); // false if the groups haven't changed
+  const IntList &requestedGridObjs(); // the requested groups
+  const BigRealList &gridObjForces(); // the corresponding forces on groups
+  bool requestedTotalForces() { return totalForceRequested; }
+
+  /* sets changedAtoms and changedForces to false again */
+  void clearChanged();
+  virtual ~GlobalMaster(); // necessary for abstract classes '-P
+
+  void check() const; // dies if there are problems with the rep invariants
+
+  void setLattice(const Lattice *lat) { lattice = lat; }
+
+  friend class colvarproxy_namd; // Temporary until refactoring is complete
```

Comment:
I think you can re-implement the functions required by `colvarproxy_namd` in `GlobalMasterColvars` as public, so that you can bypass the protected functions here.

Reply:
Totally agree. This is just a way to make it work short-term